feat: Nested NetworkBehaviour Synchronization [MTT-5024]

NoelStephensUnity opened the pull request:
This refactor provides users a way for synchronizing nested NetworkBehaviours.
It also fixes an issue with the serialization pipeline where if a NetworkObject fails to synchronize it will not halt the entire synchronization process.  

- This also resolves the issue where nested NetworkTransforms were not being synchronized.
- When variable length safety is disabled it does not use an unsigned short to define the size written per NetworkVariable.

[MTT-5024](https://jira.unity3d.com/browse/MTT-5024)
Associated reports of this issue:
https://discordapp.com/channels/449263083769036810/1033497351475699712/1033497351475699712
https://discordapp.com/channels/449263083769036810/1034802372880375828/1034802372880375828

## Changelog
- Added: Protected method `NetworkBehaviour.OnSynchronize` which is invoked during the initial `NetworkObject` synchronization process. This provides users the ability to include custom serialization information that will be applied to the `NetworkBehaviour` prior to the `NetworkObject` being spawned.
- Fixed: Issue where `NetworkTransform` components nested under a parent with a `NetworkObect` component  (i.e. network prefab) would not have their associated `GameObject`'s transform synchronized.
- Fixed: Issue where `NetworkObject`s that failed to instantiate could cause the entire synchronization pipeline to be disrupted/halted for a connecting client.
- Fixed: Issue where in-scene placed `NetworkObject`s nested under a `GameObject` would be added to the orphaned children list causing continual console warning log messages.
- Changed: When `NetworkConfig.EnsureNetworkVariableLengthSafety` is disabled `NetworkVariable` fields do not write the additional `ushort` size value (_which helps to reduce the total synchronization message size_), but when enabled it still writes the additional `ushort` value.

## Testing and Documentation
- Includes integration test updates and additions

NoelStephensUnity committed:
fix and refactor

This includes the changes to provide users with a way to synchronize NetworkBehaviours with custom data prior to their associated NetworkObject is spawned.
This also includes some fixes that allows for NetworkObjects to fail, NetworkVariables to fail, and NetworkBehaviour synchronization to fail without impacting the rest of the synchronization process.

NoelStephensUnity committed:
test

An updated version of this test that includes additional tests related to the refactoring.

This test was not properly testing the entire serialization pipeline and it only tested for one NetworkVariable.  If you added another NetworkVariable to the NetworkBehaviourWithNetworkVariables component the synchronization process would halt.  There was also a bug in SceneEventData that would occur if a NetworkObject failed to be instantiated. The updates included resolve all of these issues for the original intention behind the test.

NoelStephensUnity committed:
update

This updates NetworkTransform so that it will properly synchronize when placed on nested NetworkBehaviours (i.e. their GameObject has no NetworkObject component).

NoelStephensUnity committed:
test manual

This include a nested NetworkTransform manual test to visually validate the nested NetworkTransform update.

NoelStephensUnity committed:
fix

Minor fix for in-scene placed parenting under a non-NetworkObject to prevent from being added to the orphaned child list.

NoelStephensUnity committed:
test update

minor camera related updates to the nested NetworkTransform manual test.

NoelStephensUnity committed:
style

finishing a comment

NoelStephensUnity committed:
style

Updated comments, formatting, and organizational related issues.

NoelStephensUnity committed:
fix

compilation issue.

jeffreyrainy commented on com.unity.netcode.gameobjects/Runtime/Core/NetworkObject.cs:
```diff
@@ -1325,26 +1397,44 @@ internal SceneObject GetMessageSceneObject(ulong targetClientId)
+                try
+                {
+                    // If we failed to load this NetworkObject, then skip past the Network Variable and (if any) synchronization data
+                    reader.ReadValueSafe(out ushort networkBehaviourSynchronizationData);
```
`networkBehaviourSynchronizationData` seems to be a length. Can we call it `networkBehaviourSynchronizationDataLength` ?

NoelStephensUnity replied:
Yeah, good call. I tried to get all of the position offset names to be more meaningful...and that adjustment will indeed make more sense.

NoelStephensUnity committed:
style

variable name and comment adjustments.

NoelStephensUnity committed:
fix

Yamato really doesn't like that conditional even though it compiles just fine.

jeffreyrainy commented on com.unity.netcode.gameobjects/Runtime/Core/NetworkBehaviour.cs:
```diff
@@ -771,27 +781,43 @@ internal void WriteNetworkVariableData(FastBufferWriter writer, ulong targetClie
                 }
-                else
+                else // Only if EnsureNetworkVariableLengthSafety, otherwise just skip
+                if (NetworkManager.NetworkConfig.EnsureNetworkVariableLengthSafety)
```
I feel uneasy about relying on both sides to read/write the same stuff. When, inevitably, we find a race condition and one side still thinks a client can read a given variable and the other think it cannot, are we not writing and reading differently, thus causing a very-hard-to-debug issue ?

NoelStephensUnity replied:
We can discuss.
This code is only invoked when synchronizing (i.e. late join and newly spawned)  `NetworkObject`s and the permissions are already set prior to invoking this code.

NoelStephensUnity committed:
style

white space removal

NoelStephensUnity committed:
Merge branch 'develop' into fix/nested-networktransform-synchronization

NoelStephensUnity committed:
update

MTT-5024
updating changelog with the addition, fixes, and change

NoelStephensUnity committed:
test manual

Minor updates to some manual test assets.

NoelStephensUnity committed:
Merge branch 'develop' into fix/nested-networktransform-synchronization

NoelStephensUnity committed:
fix - merge issues

fixing some minor merge issues with the byte packing updates.

NoelStephensUnity committed:
style

removing whitespaces that somehow made their way in when merging.

NoelStephensUnity committed:
test

Renamed NetworkObjectSceneSerializationTests to NetworkObjectSynchronizationTests.
Added a more robust/wider range of tests, also added running host or server.
Fixed an issue with late joined clients not registering all players in NetcodeIntegrationTest.

NoelStephensUnity committed:
test

Integration test to validate nested network transforms synchronize properly with the nested GameObjects' transforms.

NoelStephensUnity committed:
style

removing unused const string value.

NoelStephensUnity committed:
Merge branch 'develop' into fix/nested-networktransform-synchronization

NoelStephensUnity committed:
Merge branch 'develop' into fix/nested-networktransform-synchronization

NoelStephensUnity committed:
style

Updated and added comments for readability
Removed whitespace and a local network variable that was not needed.

NoelStephensUnity committed:
Merge branch 'develop' into fix/nested-networktransform-synchronization

NoelStephensUnity committed:
style

one last minor adjustment to a comment for better clarity.

NoelStephensUnity committed:
style

>.<
Spelling issue.

NoelStephensUnity committed:
style

Removing unused using directive.

jeffreyrainy commented on com.unity.netcode.gameobjects/Components/NetworkTransform.cs:
```diff
@@ -451,6 +451,33 @@ internal NetworkTransformState GetLastSentState()
+        protected override void OnSynchronize<T>(ref BufferSerializer<T> serializer)
+        {
+            // We don't need to synchronize NetworkTransforms that are on the same
+            // GameObject as the NetworkObject.
```
Wait, now I'm confused. What is `OnSynchronize`, and why does it involve multiple objects? Why is it not just "This NetworkTransform is on one object, we synchronize the position of this object" ?

NoelStephensUnity replied:
Marking this resolved after our 1:1 huddle.

NoelStephensUnity committed:
Merge branch 'develop' into fix/nested-networktransform-synchronization

NoelStephensUnity committed:
test

Adding a basic OnSynchronize test.

NoelStephensUnity committed:
style

updated comments

jeffreyrainy reviewed (approved):
For the record, I have reservation around the direction and the NetworkObject nested restriction. I feel we walked pretty far away from the MLAPI design relying on NetworkVariables and RPC in a pretty uniform fashion.

That said, the PR shows thorough code, properly tested and addresses specific shortcomings. So, let's go with this, but consider marking a bit more clearly "OnSynchronize is optional and only meant to address complex cases where NetworkVariables and RPC fall short"

NoelStephensUnity committed:
Merge branch 'develop' into fix/nested-networktransform-synchronization

NoelStephensUnity committed:
style

updating public API xml documentation for NetworkBehaviour.OnSynchronize
